code/game/objects/structures/barrels/cask_rack.dm

```diff
@@ -0,0 +1,146 @@
+	material            = /decl/material/solid/organic/wood
+	color               = /decl/material/solid/organic/wood::color
+	material_alteration = MAT_FLAG_ALTERATION_COLOR | MAT_FLAG_ALTERATION_NAME | MAT_FLAG_ALTERATION_DESC
+	var/max_stack = 1
+
+/obj/structure/cask_rack/Initialize(ml, _mat, _reinf_mat)
+	. = ..()
+	if(isturf(loc))
+		for(var/atom/movable/stackable in loc)
+			if(try_stack_barrel(stackable) && length(contents) >= max_stack)
+				return
+
+/obj/structure/cask_rack/examine(mob/user, distance, infix, suffix)
+	. = ..()
+	if(length(contents))
+		to_chat(user, SPAN_NOTICE("It contains [english_list(contents)]."))
+
+/obj/structure/cask_rack/handle_mouse_drop(atom/over, mob/user, params)
+	if(isturf(over) && user.Adjacent(over) && Adjacent(over) && try_unstack_barrel(target = over, user = user))
+		return
+	return ..()
+
+/obj/structure/cask_rack/receive_mouse_drop(atom/dropping, mob/user, params)
+	. = ..()
+	if(!. && user.Adjacent(src) && dropping.Adjacent(src) && user.Adjacent(dropping))
+		return try_stack_barrel(dropping, user)
+
+/obj/structure/cask_rack/on_update_icon()
+	. = ..()
+	if(length(contents))
+		// Workaround for base color getting applied to vis_contents.
+		var/base_color = get_color()
+		color = null
+		var/image/I = image(icon, icon_state)
+		I.color = base_color
+		add_overlay(I)
+		// Reposition/update our contents.
+		var/i = 0
+		var/list/stackable_types = get_stackable_barrel_types()
+		for(var/atom/movable/barrel in contents)
+			if(is_type_in_list(barrel, stackable_types))
+				i++
+				adjust_barrel_offsets(barrel, i)
+	else
+		color = get_color()
+	compile_overlays() // Avoid wonky flickering on contents changes
+
+/obj/structure/cask_rack/proc/adjust_barrel_offsets(atom/movable/barrel, barrel_position)
+	barrel.reset_offsets(anim_time = 0)
+	barrel.vis_flags |= (VIS_INHERIT_LAYER | VIS_INHERIT_PLANE)
+
+/obj/structure/cask_rack/Entered(atom/movable/AM, atom/old_loc)
+	. = ..()
+	if(istype(AM) && !QDELETED(AM) && is_type_in_list(AM, get_stackable_barrel_types()))
+		vis_contents |= AM
+		recalculate_barrel_values()
+
+/obj/structure/cask_rack/Exited(atom/movable/AM, atom/new_loc)
+	. = ..()
+	if(istype(AM) && is_type_in_list(AM, get_stackable_barrel_types()))
+		vis_contents -= AM
+		AM.vis_flags = initial(AM.vis_flags)
+		AM.reset_offsets(anim_time = 0)
+		recalculate_barrel_values()
+
+/obj/structure/cask_rack/proc/recalculate_barrel_values()
+	if(length(contents))
+		density     =  TRUE
+		anchored    =  TRUE
+		obj_flags  &= ~OBJ_FLAG_ANCHORABLE
+		atom_flags |=  ATOM_FLAG_CLIMBABLE
+	else
+		density     =  FALSE
+		obj_flags  |=  OBJ_FLAG_ANCHORABLE
+		atom_flags &= ~ATOM_FLAG_CLIMBABLE
+	update_icon()
+
+/obj/structure/cask_rack/proc/try_unstack_barrel(atom/movable/barrel, turf/target, mob/user)
+	if(!loc)
+		return FALSE
+	if(!barrel)
+		if(!length(contents))
+			to_chat(user, SPAN_WARNING("\The [src] has nothing stacked on it."))
+			return FALSE
+		barrel = contents[length(contents)]
```

Comment:
a little worried about the safety of this, but I guess it doesn't make sense for anything else to ever enter it?